proposals/0000-modifiers.rst

```diff
@@ -0,0 +1,233 @@
+a function that must consume its input ``Int`` precisely once. The fact that
+the function is linear is indicated by the ``%1`` before the ``->``.
+
+A separate proposal `#242`_ describes how the matchability of a function will
+be written postfix, with a ``@``. (The precise meaning of "matchability" is
+irrelevant here.) So, if someone wanted to specify both the multiplicity and
+the matchability of a function, they would write e.g. ``Int %Many -> @Unmatchable Bool``.
+The problme is that we can easily imagine yet more decorations one would want
+to put on an arrow. With multiplicity coming before and matchability afterwards,
+there is now no more room for other modifiers (such as injectivity, or visibility,
+or higher-order role, or others we haven't imagined yet).
+
+In addition, GHC *already* supports some modifier-like constructs, but uses
+a different syntax. Specifically, the term-level operators ``inline`` and
+``oneShot`` are given types like the identity function, but they are really
+modifiers on the variable that comes afterwards. We might imagine making
+their special status more explicit by writing them ``%inline`` or ``%oneShot``.
+
+Proposed Change Specification
+-----------------------------
+
+1. Introduce a new extension ``-XModifiers``.
+
+2. With ``-XModifiers``, introduce modifier syntax on arrows as follows (cf.
+   the Haskell 2010 Report for all BNF syntax, and recall that we use braces
+   to denote "zero or more")::
+
+     type     ::= btype [ {modifier} -> type ]
+     prefix%  ::= '%'    -- only in prefix position
+     modifier ::= prefix% atype
+
+3. With ``-XModifiers``, introduce modifier syntax on types as follows::
+
+     btype    ::= {modifier} atype | btype atype
+
+4. With ``-XModifiers``, introduce modifier syntax on the term level as follows::
+
+     fexp     ::= {modifier} aexp | fexp aexp
+
+5. With ``-XModifiers``, introduce postfix modifier syntax on patterns as follows::
+
+     lpat     ::= apat {modifier}
+
+6. With ``-XModifiers``, introduce postfix modifier syntax on record field declarations as follows::
+     
+     fielddecl ::= vars '::' (type | '!' atype) {modifier}
+     
+7. Reserve the use of ``%`` in a prefix occurrence to be used only for modifiers;
+   though this proposal does not do so, we can imagine extending the modifier syntax
+   to apply to further syntactic situations (e.g. term-level operators, declarations,
+   import lists, etc.). The one exception is the syntax ``%1`` for a linear function,
+   which continues to be allowed.
+
+8. Introduce a new type-level constant ``Modifier :: Type -> Constraint``, exported
+   from ``GHC.Exts``.
+
+9. Let the constraint ``Modifier Multiplicity`` be satisfiable; let no other
+   ``Modifier`` constraint be satisfiable.
+
+10. During constraint generation, let an occurrence ``%(ty)``, where ``ty :: ki``,
+    emit a constraint ``Modifier ki``.
+
+11. A modifier of type ``Multiplicity`` changes the multiplicity of the following arrow,
```

Comment:
The idea of constraint resolution changing the type of a neighboring syntactic element feels really strange to me. And it's very unclear how we would extend this to user-defined modifiers without a user-definable semantics for "running" a modifier, which I think is a must for a feature like this (see other existing mechanisms like `ANN` pragmas, the various plugins we support, even template haskell and quasiquoters). I think I'd be happier with this proposal if it specified a user-definable mechanism and then showed how we could use it to solve the multiplicity/matchability problem.

Reply:
Yes, I've grown uncomfortable about this, too. Modification of the proposal to appear shortly.